app/javascript/components/tabs.jsx

```diff
@@ -20,35 +23,116 @@ const OtherWidgetText = styled.div`
   cursor: pointer;
 `;
 
-export const TabBar = () => (
-  <svg width="568" height="4">
-    <rect width="568" height="1" fill={colors.white} opacity="0.2" />
-  </svg>
-);
+const TabButton = styled.div`
+  outline: none;
+`;
+
+export class Tab extends React.Component {
+  static propTypes = {
+    selected: PropTypes.bool.isRequired,
+    text: PropTypes.string.isRequired,
+    widgetId: PropTypes.number.isRequired,
+    totalTime: PropTypes.number.isRequired,
+    tabDown: PropTypes.func.isRequired,
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      timeBarProgress: this.props.totalTime,
+      maxTime: this.props.totalTime,
+    };
+    this.timeRemaining = this.props.totalTime;
+    this.isPaused = false;
+  }
+
+  componentDidMount() {
+    if (this.props.selected) {
+      this.timer();
+    }
+  }
+
+  componentDidUpdate(prevProps) {
+    const isSelected = this.props.selected;
+    if (isSelected !== prevProps.selected) {
+      if (isSelected && !this.intervalId) {
+        this.timer();
+      } else if (!isSelected) {
+        clearInterval(this.intervalId);
+        this.intervalId = null;
+        this.timeRemaining = this.state.maxTime;
+        if (this.isPaused) {
+          this.togglePause();
+        }
+      }
+    }
+  }
+
+  componentWillUnmount() {
+    if (this.intervalId) {
+      clearInterval(this.intervalId);
+    }
+  }
+
+  timer() {
+    const max = this.state.maxTime;
+    this.intervalId = setInterval(() => {
+      if (this.props.selected) {
+        if (!this.isPaused) {
+          this.timeRemaining -= 50;
+        } else {
+          this.timeRemaining = max;
+        }
+        if (this.timeRemaining <= 0) {
+          this.timeRemaining = max;
+          this.props.tabDown();
+        }
+      } else {
+        this.timeRemaining = max;
+        if (this.isPaused) {
+          this.togglePause();
+        }
+      }
+      this.setState({ timeBarProgress: this.timeRemaining });
+    }, 50);
+  }
+
+  togglePause() {
+    this.isPaused = !this.isPaused;
+  }
 
-export const Tab = props => {
-  if (props.selected) {
+  render() {
+    const isSelected = this.props.selected;
+    let widgetBar;
+    if (isSelected) {
+      widgetBar = (
+        <span>
+          <CurrentTabText id={`widget_${this.props.widgetId}`}>
+            {this.props.text}
+          </CurrentTabText>
+          <CurrentTabBar
+            time={this.state.timeBarProgress}
+            max={this.state.maxTime}
+          />
+        </span>
+      );
+    } else {
+      widgetBar = (
+        <span>
+          <OtherTabText id={`widget_${this.props.widgetId}`}>
+            {this.props.text}
+          </OtherTabText>
+          <OtherTabBar />
+        </span>
+      );
+    }
     return (
-      <span>
-        <CurrentWidgetText id={`widget_${props.widgetId}`}>
-          {props.text}
-        </CurrentWidgetText>
-        <TabBar />
-      </span>
+      // eslint-disable-next-line
```

Comment:
I'd just move that `onKeyDown` inline fn expression into a class method. It's a little more readable to me when it's not inline in the render method, and it's arguably a good practice (since creating a new method here will invalidate this part of the render tree in React's eyes).

```render
class Tabs {
  handleKeyDown = ({ keyCode }) => ...;

  render() {
    <TabButton onKeyDown={} />
  }
```